exporter/humioexporter/traces_exporter_test.go

```diff
@@ -16,30 +16,467 @@ package humioexporter
 
 	"github.com/stretchr/testify/assert"
 	"github.com/stretchr/testify/require"
-	"go.opentelemetry.io/collector/config"
 	"go.opentelemetry.io/collector/consumer/consumererror"
 	"go.opentelemetry.io/collector/consumer/pdata"
+	"go.opentelemetry.io/collector/translator/conventions"
 	"go.uber.org/zap"
 )
 
+func createSpanID(stringVal string) [8]byte {
+	var id [8]byte
+	b, _ := hex.DecodeString(stringVal)
+	copy(id[:], b)
+	return id
+}
+
+func createTraceID(stringVal string) [16]byte {
+	var id [16]byte
+	b, _ := hex.DecodeString(stringVal)
+	copy(id[:], b)
+	return id
+}
+
+// Implement a mock of the client interface for testing
+type clientMock struct {
+	response func() error
+}
+
+func (m *clientMock) sendUnstructuredEvents(ctx context.Context, evts []*HumioUnstructuredEvents) error {
+	return m.response()
+}
+
+func (m *clientMock) sendStructuredEvents(ctx context.Context, evts []*HumioStructuredEvents) error {
+	return m.response()
+}
+
 func TestPushTraceData(t *testing.T) {
 	// Arrange
-	exp := newTracesExporter(&Config{ExporterSettings: config.NewExporterSettings(typeStr)}, zap.NewNop(), nil)
+	testCases := []struct {
+		desc     string
+		client   exporterClient
+		wantErr  bool
+		wantPerm bool
+	}{
+		{
+			desc: "Valid request",
+			client: &clientMock{
+				response: func() error {
+					return nil
+				},
+			},
+			wantErr:  false,
+			wantPerm: false,
+		},
+		{
+			desc: "Forwards transient errors",
+			client: &clientMock{
+				response: func() error {
+					return errors.New("Error")
+				},
+			},
+			wantErr:  true,
+			wantPerm: false,
+		},
+		{
+			desc: "Forwards permanent errors",
+			client: &clientMock{
+				response: func() error {
+					return consumererror.Permanent(errors.New("Error"))
+				},
+			},
+			wantErr:  true,
+			wantPerm: true,
+		},
+	}
 
 	// Act
-	err := exp.pushTraceData(context.Background(), pdata.NewTraces())
+	for _, tC := range testCases {
+		t.Run(tC.desc, func(t *testing.T) {
+			exp := newTracesExporter(&Config{}, zap.NewNop(), tC.client)
+			err := exp.pushTraceData(context.Background(), pdata.NewTraces())
+
+			// Assert
+			if (err != nil) != tC.wantErr {
+				t.Errorf("pushTraceData() error = %v, wantErr %v", err, tC.wantErr)
+			}
+
+			if consumererror.IsPermanent(err) != tC.wantPerm {
+				t.Errorf("pushTraceData() permanent = %v, wantPerm %v",
+					consumererror.IsPermanent(err), tC.wantPerm)
+			}
+		})
+	}
+}
+
+func TestPushTraceData_PermanentOnCompleteFailure(t *testing.T) {
+	// Arrange
+	// We do not export spans with missing service names, so this span should
+	// fail exporting
+	traces := pdata.NewTraces()
+	traces.ResourceSpans().Resize(1)
+	traces.ResourceSpans().At(0).InstrumentationLibrarySpans().Resize(1)
+	traces.ResourceSpans().At(0).InstrumentationLibrarySpans().At(0).Spans().Resize(1)
+
+	exp := newTracesExporter(&Config{}, zap.NewNop(), &clientMock{})
+
+	// Act
+	err := exp.pushTraceData(context.Background(), traces)
 
 	// Assert
+	require.Error(t, err)
```

Comment:
Make sure error message returned for missing service name is very prescriptive since export is being failed. You may also want to log the error in the converter where the service name is extracted.